pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,529 @@
+            coeff *= _get_kernel_expression(kernel.global_scaling_const,
+                    kernel_arguments)
+            coeff /= _get_kernel_expression(base_kernel.global_scaling_const,
+                    kernel_arguments)
+            result.append((mis[i], sympy_conv(coeff)))
+            logger.debug("  + %s.diff(%s)*%s", base_kernel, mis[i], coeff)
+        else:
+            const = sympy_conv(coeff * _get_kernel_expression(
+                kernel.global_scaling_const, kernel_arguments))
+    logger.debug("  + %s", const)
+    return (const, result)
+
+
+@memoize_on_first_arg
+def _get_base_kernel_matrix(base_kernel: ExpressionKernel,
+        hashable_kernel_arguments: Tuple[Tuple[Text, Any]],
+        order: Optional[int] = None, retries: int = 3) \
+        -> Tuple[Tuple[sym.Matrix, sym.Matrix, List[Tuple[int, int]]],
+            np.ndarray, List[Tuple[int]]]:
+    dim = base_kernel.dim
+
+    pde = base_kernel.get_pde_as_diff_op()
+    if order is None:
+        order = pde.order
+
+    if order > pde.order:
+        raise RuntimeError(f"order ({order}) cannot be greater than the order"
+                         f"of the PDE ({pde.order}) yet.")
+
+    mis = sorted(gnitstam(order, dim), key=sum)
+    # (-1, -1, -1) represent a constant
+    mis.append((-1, -1, -1))
+
+    if order == pde.order:
+        pde_mis = [ident.mi for eq in pde.eqs for ident in eq.keys()]
+        pde_mis = [mi for mi in pde_mis if sum(mi) == order]
+        logger.debug(f"Removing {pde_mis[-1]} to avoid linear dependent mis")
+        mis.remove(pde_mis[-1])
+
+    rand = np.random.randint(1, 10**15, (dim, len(mis)))
+    rand = rand.astype(object)
+    for i in range(rand.shape[0]):
+        for j in range(rand.shape[1]):
+            rand[i, j] = sym.sympify(rand[i, j])/10**15
+    sym_vec = make_sym_vector("d", dim)
+
+    base_expr = _get_kernel_expression(base_kernel.expression,
+        dict(hashable_kernel_arguments))
+
+    mat = []
+    for rand_vec_idx in range(rand.shape[1]):
+        row = []
+        for mi in mis[:-1]:
+            expr = base_expr
+            for var_idx, nderivs in enumerate(mi):
+                if nderivs == 0:
+                    continue
+                expr = expr.diff(sym_vec[var_idx], nderivs)
+            replace_dict = dict(
+                (k, v) for k, v in zip(sym_vec, rand[:, rand_vec_idx])
+            )
+            eval_expr = evalf(expr.xreplace(replace_dict))
+            row.append(eval_expr)
+        row.append(1)
+        mat.append(row)
+
+    mat = sym.Matrix(mat)
+    failed = False
+    try:
+        L, U, perm = mat.LUdecomposition()
+    except RuntimeError:
+        # symengine throws an error when rank deficient
+        # and sympy returns U with last row zero
+        failed = True
+
+    if not sym.USE_SYMENGINE and all(expr == 0 for expr in U[-1, :]):
+        failed = True
+
+    if failed:
+        if retries == 0:
+            raise RuntimeError("Failed to find a base kernel")
```

Comment:
Maybe make a special purpose error for this? `BaseKernelNotFoundError` or something?

Reply:
Better error message in 106816f78e0a7f572c6f796999a6ab0133c4c858